pinata/pinata.go

```diff
@@ -0,0 +1,225 @@
+package pinata
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/opensaucerer/bifrost/shared/errors"
+	"github.com/opensaucerer/bifrost/shared/types"
+	"io"
+	"mime/multipart"
+	"net/http"
+	"os"
+	"path/filepath"
+	 "encoding/json"
+)
+
+type MetaData struct {
+    Name   string      `json:"name"`
+}
+
+type SuccessResponse struct {
+	 IpfsHash   string `json:"IpfsHash"`
+	 Timestamp 	string `json:"Timestamp"`
+	 PinSize 		int64 `json:"PinSize"`
+}
+
+/*
+	Upload File to pinata IPFS
+*/
+
+func (p PinataIPFSStorage) UploadFile(path, filename string, options map[string]interface{}) (*types.UploadedFile, error) {
+
+  // validate JWT	
+	success,err := p.TestAuthentication()
+	if err != nil {
+		return nil, &errors.BifrostError{
+			Err:       fmt.Errorf("problem authenticating: %v", err),
+			ErrorCode: errors.ErrBadRequest,
+		}
+	}
+	if !success {
+		return nil, &errors.BifrostError{
+			Err:       fmt.Errorf("authentication failed"),
+			ErrorCode: errors.ErrBadRequest,
+		}
+	}
+
+	if _, err := os.Stat(path); os.IsNotExist(err) {
+		return nil, &errors.BifrostError{
+			Err:       fmt.Errorf("file does not exist: %s", path),
+			ErrorCode: errors.ErrBadRequest,
+		}
+	}
+
+	url := "https://api.pinata.cloud/pinning/pinFileToIPFS"
+	method := "POST"
+
+	payload := &bytes.Buffer{} // buffer writer
+	writer := multipart.NewWriter(payload)
+	file, fileOpenError := os.Open(path)
+
+	if fileOpenError != nil {
+		return nil, &errors.BifrostError{
+			Err:       fileOpenError,
+			ErrorCode: errors.ErrFileOperationFailed,
+		}
+	}
+
+	part1, err := writer.CreateFormFile("file", filepath.Base(path))
+	if err != nil {
+		return nil, &errors.BifrostError{
+			Err:       fmt.Errorf("file does not exist: %s", path),
+			ErrorCode: errors.ErrBadRequest,
+		}
+	}
+
+	_, errFile1 := io.Copy(part1, file)
+	if errFile1 != nil {
+		return nil, &errors.BifrostError{
+			Err:       fmt.Errorf("failed to copy multipart data"),
+			ErrorCode: errors.ErrBadRequest,
+		}
+	}
+
+	defer file.Close()
```

Comment:
We should defer the closing of a file the moment we open it...if an error occurs before getting to this `defer` then the file will never be closed and memory locations won't be freed.

Reply:
that's absolutely correct.